File: 032-base64/base64.java
import java.io.*;

public class base64 {
  private static final int BUF_SIZE = 4096;

  private static char[] BASE64_CHARS = {
    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N',
    'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b',
    'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p',
    'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '0', '1', '2', '3',
    '4', '5', '6', '7', '8', '9', '+', '/'
  };

  private static int encode(int n, byte[] input, char[] output, Writer outputWriter) throws IOException {
    int chunks = n / 3;
    int oidx = 0;
    for (int c = 0; c < chunks; c++) {
      int b1 = input[c * 3 + 0] & 0xff;
      int b2 = input[c * 3 + 1] & 0xff;
      int b3 = input[c * 3 + 2] & 0xff;
      output[oidx++] = BASE64_CHARS[(b1 >> 2) & 0x3f];
      output[oidx++] = BASE64_CHARS[((b1 << 4) & 0x30) | ((b2 >> 4) & 0x0f)];
      output[oidx++] = BASE64_CHARS[((b2 << 2) & 0x3c) | ((b3 >> 6) & 0x03)];
      output[oidx++] = BASE64_CHARS[(b3 & 0x3f)];
    }

    outputWriter.write(output, 0, oidx);
    return chunks * 3;
  }

  public static void encodePadding(int n, byte[] input, char[] output, Writer outputWriter) throws IOException {
    int oidx = 0;
    if (n == 1) {
      int b1 = input[0] & 0xff;
      output[oidx++] = BASE64_CHARS[(b1 >> 2) & 0x3f];
      output[oidx++] = BASE64_CHARS[(b1 << 4) & 0x30];
      output[oidx++] = '=';
      output[oidx++] = '=';
    } else if (n == 2) {
      int b1 = input[0] & 0xff;
      int b2 = input[1] & 0xff;
      output[oidx++] = BASE64_CHARS[(b1 >> 2) & 0x3f];
      output[oidx++] = BASE64_CHARS[((b1 << 4) & 0x30) | ((b2 >> 4) & 0x0f)];
      output[oidx++] = BASE64_CHARS[(b2 << 2) & 0x3c];
      output[oidx++] = '=';
    }
    outputWriter.write(output, 0, oidx);
    return;
  }

  public static void main(String[] args) throws IOException {
    var stdin = System.in;
    var stdout = new BufferedWriter(new OutputStreamWriter(System.out));

    var inputbuf = new byte[BUF_SIZE];
    var outputbuf = new char[BUF_SIZE + BUF_SIZE / 2];
    int rem = 0;
    while (true) {
      int n = stdin.read(inputbuf, rem, inputbuf.length - rem);
      if (n <= 0)
        break;

      int m = encode(n + rem, inputbuf, outputbuf, stdout);
      rem = (n + rem) % 3;
      for (int i = 0; i < rem; i++)
        inputbuf[i] = inputbuf[m+i];
    }

    if (rem != 0)
      encodePadding(rem, inputbuf, outputbuf, stdout);

    stdout.close();
  }
}

File: 032-base64/base64.c
#include <assert.h>
#include <errno.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
 * A rather naive base64 encoding impl.! More sophisticated stuff to come.
 */

static const size_t BUFSIZE = 4096;
static const uint8_t base64_chars[64] =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

/* Returns the number of bytes treated. If done is true, that is the last
 * chunck of that to be encoded, and all n bytes should be handled, as well as
 * padding characters written if necessary.
 */
static size_t base64_encode(size_t n, uint8_t input[n], bool done, FILE *dst) {
  size_t chunks = n / 3, idx = 0;
  uint8_t output[BUFSIZE + BUFSIZE / 2];
  for (size_t c = 0; c < chunks; c++) {
    uint32_t b1 = input[c * 3 + 0] & 0xff, b2 = input[c * 3 + 1] & 0xff,
             b3 = input[c * 3 + 2] & 0xff;
    uint8_t c1 = base64_chars[(b1 >> 2) & 0x3f];
    uint8_t c2 = base64_chars[((b1 << 4) & 0x30) | ((b2 >> 4) & 0x0f)];
    uint8_t c3 = base64_chars[((b2 << 2) & 0x3c) | ((b3 >> 6) & 0x03)];
    uint8_t c4 = base64_chars[((b3 & 0x3f))];
    output[idx++] = c1;
    output[idx++] = c2;
    output[idx++] = c3;
    output[idx++] = c4;
  }

  size_t rem = n % 3;
  if (!done || rem == 0) {
    fwrite(output, sizeof(uint8_t), idx, dst);
    return chunks * 3;
  }

  if (rem == 1) {
    uint32_t b1 = input[chunks * 3 + 0] & 0xff;
    output[idx++] = base64_chars[(b1 >> 2) & 0x3f];
    output[idx++] = base64_chars[(b1 << 4) & 0x30];
    output[idx++] = '=';
    output[idx++] = '=';
  } else if (rem == 2) {
    uint32_t b1 = input[chunks * 3 + 0] & 0xff;
    uint32_t b2 = input[chunks * 3 + 1] & 0xff;
    output[idx++] = base64_chars[(b1 >> 2) & 0x3f];
    output[idx++] = base64_chars[((b1 << 4) & 0x30) | ((b2 >> 4) & 0x0f)];
    output[idx++] = base64_chars[(b2 << 2) & 0x3c];
    output[idx++] = '=';
  }

  fwrite(output, sizeof(uint8_t), idx, dst);
  return n;
}

int main(int argc, const char *argv[]) {
  if (argc > 1) {
    fprintf(stderr,
            "%s: A base64 encoder. Reads (binary data) from stdin, "
            "writes (base64) to stdout.\n",
            argv[0]);
    return EXIT_FAILURE;
  }

  uint8_t input[BUFSIZE];
  size_t start_offset = 0;
  size_t n = -1ul;
  for (;;) {
    n = fread(&input[start_offset], sizeof(uint8_t), BUFSIZE - start_offset,
              stdin);
    if (n <= 0) {
      if (ferror(stdin)) {
        fprintf(stderr, "%s: failed to read from stdin: %s\n", argv[0],
                strerror(errno));
        return EXIT_FAILURE;
      }

      if (feof(stdin))
        break;
    }

    bool done = (n < BUFSIZE - start_offset) && feof(stdin);

    assert(n <= BUFSIZE - start_offset);
    /* done could be initialised with feof() if n < BUFSIZE - start_offset. */
    size_t m = base64_encode(n + start_offset, &input[0], done, stdout);
    if (m == n + start_offset) {
      start_offset = 0;
    } else {
      start_offset = (n + start_offset) - m;
      memmove(&input[0], &input[m], start_offset);
    }
  }

  if (start_offset != 0)
    base64_encode(start_offset, &input[0], true, stdout);

  return EXIT_SUCCESS;
}
